feat(common): expose component instance in NgComponentOutlet

crisbeto opened the pull request:
Exposes the current instance of the component in the `NgComponentOutlet` directive.

Harpush commented:
Any chance for rendered output which passes the component ref as parameter? Mostly for class functions interactions

JeanMeche commented on packages/common/src/directives/ng_component_outlet.ts:
```diff
@@ -118,6 +118,14 @@ export class NgComponentOutlet implements OnChanges, DoCheck, OnDestroy {
+   * Will be null if no component has been rendered.
+   */
+  get componentInstance(): T | null {
+    return this._componentRef?.instance || null;
```
Any reason for not going with the nullish coalescing operator here ?

crisbeto replied:
The `instance` can't be false or 0 so it doesn't make a difference.

pkozlowski-opensource replied:
I'm with @JeanMeche -  with `??` I don't need to think / consider those cases

JoostK commented on packages/common/src/directives/ng_component_outlet.ts:
```diff
@@ -95,7 +95,7 @@ import {
 })
-export class NgComponentOutlet implements OnChanges, DoCheck, OnDestroy {
+export class NgComponentOutlet<T = unknown> implements OnChanges, DoCheck, OnDestroy {
   @Input() ngComponentOutlet: Type<any> | null = null;
```
If this were `Type<T>` then type inference would probably work from the template; that could be relevant if this directive were to also have `exportAs` to be able to access the component instance from a template reference instead of `contentChild`.

The `T` defaulting to `unknown` may be a breaking change concern in that case.

crisbeto replied:
IMO it should be `Type<T>`, I just missed to do it the first time around. I'll run a TGP to see how breaking it would be in practice. Maybe also default it to `any` instead?

crisbeto replied:
There were a couple of failures in the TGP so I think we'll have to do it in a major. I left a TODO.

pkozlowski-opensource commented on packages/common/test/directives/ng_component_outlet_spec.ts:
```diff
@@ -304,6 +304,19 @@ describe('insert/remove', () => {
     expect(fixture.nativeElement.textContent).toBe('Hello World');
   });
+
+  it('should be able to get the current component instance', waitForAsync(() => {
```
`waitForAsync`?

crisbeto committed:
feat(common): expose component instance in NgComponentOutlet

Exposes the current instance of the component in the `NgComponentOutlet` directive.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-common

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
This PR was merged into the repository by commit e4c50b3bea22ca2afba74465893c36730952f4b9.

The changes were merged into the following branches: main
